### PR TITLE
fix(quantic): handle corner-cases for ResultsPerPage

### PR DESCRIPTION
**Pull request by erocheleau**
SFINT-5954

Handle corner-cases for ResultsPerPage.

Added unit tests for those corner cases:

1. If you set the `initialChoice` with a negative value, or with not a number. There will be a console error and the component won't initialize but instead display the QuanticError.
2. If you provide a negative choicesToDisplay value, there will be a console error and the component won't initialize but instead display the QuanticError.

image

3. If the `initialChoice` provided is not part of the `choicesToDisplay`, a console warning (instead of a console error previously) will be displayed and we will default to the first value in the `choicesToDisplay`

### CODE REVIEW

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js [hunk 1]
```diff
@@ -51,7 +51,13 @@ export default class QuanticResultsPerPage extends LightningElement {
     return this._initialChoice;
   }
   set initialChoice(value) {
-    this._initialChoice = value;
+    if (Number(value) >= 0) {
+      this._initialChoice = value;
+    } else {
+      console.warn(
+        `The "initialChoice" option value "${value}" must be a number greater than 0. Defaulting to the first choices value.`
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js:58]: To stay consistent in the message logged I would either use the expression `must be a number greater than 0` always or `must be a positive number ` always, currently we use both in the same time.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js [hunk 2]
```diff
@@ -51,7 +51,13 @@ export default class QuanticResultsPerPage extends LightningElement {
     return this._initialChoice;
   }
   set initialChoice(value) {
-    this._initialChoice = value;
+    if (Number(value) >= 0) {
+      this._initialChoice = value;
+    } else {
+      console.warn(
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js:57]: Also in the consistency topic,
I find it a little bit strange that when initialChoice is <= 0 we log a warning and we default to the first value in the choices displayed without breaking the component.
But when one of the values of `choicesDisplayed ` is negative number, we log an error and we break the component by displaying the error message.

For me it seems that sometimes we are being tolerant and sometimes we are not, I would be be more comfortable if we handled every property validation the same way,  this way our library will always follow one pattern when it comes to property validation providing a better developer experience, what do you think?

Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js:57]: Well we do have a fallback if you set a wrong `initialChoice` but we possibly don't have a fallback if you set a negative or non-number in the `choicesToDisplay`.

That was my logic anyway. It seemed to make sense to "recover" from a wrong `initialChoice` but not to ignore potentially 1 or more values from the `choicesToDisplay`.

I do agree that we can also be more strict, if you pass a wrong value anywhere, we can simply error out instead.
I'll change it for consistency indeed, after your message, I think that we don't really gain anything from "recovering" from a bad value in the `initialChoice`.

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js (hunk 1 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js]: Changed it

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js (hunk 2 above)
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/quanticResultsPerPage.js]: @mmitiche let me know if it makes more sense now.

Updated the description

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js [hunk 3]
```diff
@@ -359,9 +379,65 @@ describe('c-quantic-results-per-page', () => {
...
+
+      expect(functionsMocks.buildResultsPerPage).toHaveBeenCalledTimes(1);
+      expect(functionsMocks.buildResultsPerPage).toHaveBeenCalledWith(
+        exampleEngine,
+        {
+          initialState: {
+            numberOfResults: defaultChoices[0],
+          },
+        }
+      );
```
[the remark this answers is not in the document]
Comment (mmitiche) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:403]: I would even say maybe we should not build any controller if `this.hasInitializationError` has already been set to `true`

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js [hunk 4]
```diff
@@ -359,9 +379,65 @@ describe('c-quantic-results-per-page', () => {
...
```
Comment (mmitiche) [inline packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:431]: same remark
```suggestion
```

**Review by mmitiche — APPROVED** [review]
just couple checks that I think can be removed bu the logic looks fine 👍

**Review by SimonMilord — APPROVED** [review]
Looks like a good fix to me. Just wondering maybe we can specify in the property declaration for `choicesDisplayed` that it must be a positive number

File: packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (erocheleau) [reply packages/quantic/force-app/main/default/lwc/quanticResultsPerPage/__tests__/quanticResultsPerPage.test.js:403]: wooops yea I forgot to update the test after the behavior changed xD thanks

**Review by mmitiche — APPROVED** [review]

**Review by SimonMilord — APPROVED** [review]
